sdk/eventhub/Azure.Messaging.EventHubs/design/event-processor{T}-proposal.md

```diff
@@ -0,0 +1,520 @@
+        Storage.GetCheckpointsAsync(FullyQualifiedNamespace, EventHubName, ConsumerGroup);
+    
+    protected override Task<IEnumerable<PartitionOwnership>> ListOwnershipAsync(CancellationToken cancellationToken) =>
+        Storage.GetOwnershipAsync(FullyQualifiedNamespace, EventHubName, ConsumerGroup, Identifier);
+    
+    protected override Task<IEnumerable<PartitionOwnership>> ClaimOwnershipAsync(IEnumerable<PartitionOwnership> desiredOwnership, CancellationToken cancellationToken) =>
+        Storage.TryClaimOwnershipAsync(desiredOwnership);
+}
+
+public class InMemoryStorage
+{
+    private ConcurrentDictionary<string, ConcurrentDictionary<string, EventProcessorCheckpoint>> Checkpoints { get; } = new ConcurrentDictionary<string, ConcurrentDictionary<string, EventProcessorCheckpoint>>();
+    private ConcurrentDictionary<string, ConcurrentDictionary<string, PartitionOwnership>> Ownership { get; } = new ConcurrentDictionary<string, ConcurrentDictionary<string, PartitionOwnership>>();
+
+    public Task CreateCheckpointAsync(
+        string fullyQualifiedNamespace,
+        string eventHubName,
+        string consumerGroup,
+        string partitionId,
+        long? offset,
+        long? sequenceNumber)
+    {
+        var key = CreateKey(fullyQualifiedNamespace, eventHubName, consumerGroup);
+        var checkpoints = Checkpoints.GetOrAdd(key, newKey => new ConcurrentDictionary<string, EventProcessorCheckpoint>());
+        
+        var partitionCheckpoint = checkpoints.GetOrAdd(partitionId, new EventProcessorCheckpoint
+        {
+            FullyQualifiedNamespace = fullyQualifiedNamespace,
+            EventHubName = eventHubName, 
+            ConsumerGroup = consumerGroup,
+            PartitionId = partitionId
+        });
+        
+        partitionCheckpoint.Offset = offset;
+        partitionCheckpoint.SequenceNumber = sequenceNumber;
+        
+        return Task.CompletedTask;
+    }
+
+    public Task<IEnumerable<EventProcessorCheckpoint>> GetCheckpointsAsync(
+        string fullyQualifiedNamespace,
+        string eventHubName,
+        string consumerGroup)
+    {
+        var key = CreateKey(fullyQualifiedNamespace, eventHubName, consumerGroup);
+        var checkpoints = Checkpoints.GetOrAdd(key, newKey => new ConcurrentDictionary<string, EventProcessorCheckpoint>());
+        return Task.FromResult((IEnumerable<EventProcessorCheckpoint>)checkpoints.Values);
+    }
+
+    public Task<IEnumerable<PartitionOwnership>> GetOwnershipAsync(
+        string fullyQualifiedNamespace,
+        string eventHubName,
+        string consumerGroup)
+    {
+        var key = CreateKey(fullyQualifiedNamespace, eventHubName, consumerGroup);
+        var ownership = Ownership.GetOrAdd(key, newKey => new ConcurrentDictionary<string, PartitionOwnership>());
+        return Task.FromResult(ownership.Values.Where(item => item.OwnerIdentifier != identifier));
```

Comment:
What is this `identifier`? Shouldn't we just do like in the method above?
```
return Task.FromResult((IEnumerable<PartitionOwnership>)ownership.Values);
```

Reply:
It's in the method signature of the processor today.   This is just a sample for visualization, so the absolute correctness here isn't a goal.